drivers/digital-io/max14906/max14906.c

```diff
@@ -0,0 +1,407 @@
+ * @param desc - device descriptor for the MAX14906
+ * @param addr - address of the register
+ * @param mask - bit mask of the field to be updated
+ * @param val - value of the masked field. Should be bit shifted by using
+ * 		 no_os_field_prep(mask, val)
+ * @return 0 in case of success, negative error code otherwise
+ */
+int max14906_reg_update(struct max14906_desc *desc, uint32_t addr,
+			uint32_t mask, uint32_t val)
+{
+	int ret;
+	uint32_t reg_val = 0;
+
+	ret = max14906_reg_read(desc, addr, &reg_val);
+	if (ret)
+		return ret;
+
+	reg_val &= ~mask;
+	reg_val |= mask & val;
+
+	return max14906_reg_write(desc, addr, reg_val);
+}
+
+/**
+ * @brief Read the (voltage) state of a channel (works for both input or output).
+ * @param desc - device descriptor for the MAX14906
+ * @param ch - channel index (0 based).
+ * @param val - binary value representing a channel's voltage level (0 or 1).
+ * @return 0 in case of success, negative error code otherwise
+ */
+int max14906_ch_get(struct max14906_desc *desc, uint32_t ch, uint32_t *val)
+{
+	int ret;
+
+	if (ch >= MAX14906_CHANNELS)
+		return -EINVAL;
+
+	ret = max14906_reg_read(desc, MAX14906_DOILEVEL_REG, val);
+	if (ret)
+		return ret;
+
+	*val = no_os_field_get(MAX14906_DOI_LEVEL_MASK(ch), *val);
+
+	return 0;
+}
+
+/**
+ * @brief Write the (voltage) state of a channel (only for output channels).
+ * @param desc - device descriptor for the MAX14906
+ * @param ch - channel index (0 based).
+ * @param val - binary value representing a channel's voltage level (0 or 1).
+ * @return 0 in case of success, negative error code otherwise
+ */
+int max14906_ch_set(struct max14906_desc *desc, uint32_t ch, uint32_t val)
+{
+	if (ch >= MAX14906_CHANNELS)
+		return -EINVAL;
+
+	return max14906_reg_update(desc, MAX14906_SETOUT_REG,
+				   MAX14906_HIGHO_MASK(ch), (val) ?
+				   MAX14906_HIGHO_MASK(ch) : 0);
+}
+
+/**
+ * @brief Configure a channel's function.
+ * @param desc - device descriptor for the MAX14906
+ * @param ch - channel index (0 based).
+ * @param function - channel configuration (input, output or high-z).
+ * @return 0 in case of success, negative error code otherwise
+ */
+int max14906_ch_func(struct max14906_desc *desc, uint32_t ch,
+		     enum max14906_function function)
+{
+	int ret;
+
+	if (function == MAX14906_HIGH_Z) {
+		ret = max14906_reg_update(desc, MAX14906_CONFIG_DO_REG, MAX14906_DO_MASK(ch),
+					  no_os_field_prep(MAX14906_DO_MASK(ch),
+							  MAX14906_PUSH_PULL));
+		if (ret)
+			return ret;
+
+		return max14906_reg_update(desc, MAX14906_SETOUT_REG,
+					   MAX14906_CH_DIR_MASK(ch),
+					   no_os_field_prep(MAX14906_CH_DIR_MASK(ch), 1));
+	}
+
+	if (function == MAX14906_IN) {
+		ret = max14906_reg_update(desc, MAX14906_CONFIG_DO_REG, MAX14906_DO_MASK(ch),
+					  no_os_field_prep(MAX14906_DO_MASK(ch),
+							  MAX14906_HIGH_SIDE));
+		if (ret)
+			return ret;
+	}
+
+	return max14906_reg_update(desc, MAX14906_SETOUT_REG, MAX14906_CH_DIR_MASK(ch),
```

Comment:
in case `function == MAX14906_HIGH_Z` you write write twice the same bits of MAX14906_SETOUT_REG.

Reply:
Not true.